### ISSUE TITLE
TypeError: someClass is not a constructor

### ISSUE THREAD

**Issue by MatteoGioioso** [typeorm/typeorm#5458]
**Issue type:**

[ ] question
[x] bug report
[ ] feature request
[x] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[ ] `mysql` / `mariadb`
[ ] `oracle`
[x] `postgres`
[ ] `cockroachdb`
[ ] `sqlite`
[ ] `sqljs`
[ ] `react-native`
[ ] `expo`

**TypeORM version:**

[x] `latest`
[ ] `@next`
[ ] `0.x.x` (or put your version here)

**Steps to reproduce or a small repository showing the problem:**

I have been trying for hours to generate a migrations, but seems that everything is leading to an typeorm error

this is my Entity
```
import { Entity, PrimaryGeneratedColumn, Column, CreateDateColumn, UpdateDateColumn } from "typeorm";

@Entity()
export default class CommentEntity {
  @PrimaryGeneratedColumn()
  id!: number;

  @CreateDateColumn()
  createdAt!: Date;

  @UpdateDateColumn()
  updatedAt!: Date;

  @Column()
  userNickname!: string;

  @Column()
  content!: string;
}
```

this is my ormconfig.json

```
{
   "type": "postgres",
   "host": "localhost",
   "port": 5432,
   "username": "postgres",
   "password": "",
   "database": "postgres",
   "logging": false,
   "entities": [
      "modules/comments/CommentEntity.ts"
   ],
   "migrations": [
      "migrations/*.ts"
   ],
   "cli": {
      "migrationsDir": "migrations"
   }
}
```

and then I run
```
./node_modules/.bin/ts-node ./node_modules/typeorm/cli.js migration:generate -n Comment
```

I have tried for a full text search on my IDE for `someClass` because it looks like some kind of placeholder for testing that someone forgot in the code, but there was nothing.

Thanks

**Comment by DimosthenisK** [comment]
I have exactly the same issue in mysql, I am looking at it right now.. I assume that something is wrong with my credentials.

**Comment by DimosthenisK** [comment]
I KNEW I had seen that before. Take a look at my previous issue: https://github.com/typeorm/typeorm/issues/5079

You have a file inside your migrations dir that isn't a TS file or doesn't match your migration file glob pattern

**Comment by MatteoGioioso** [comment]
Oh, I see. Thanks for finding that out I must have missed it. I will check it again.

Thanks

**Comment by kaushiksamanta** [comment]
This is the correct way to initialize migrations and entities:-
(https://github.com/nrwl/nx/issues/803)

```import { TypeOrmModuleOptions } from '@nestjs/typeorm';
import * as entities from './src/entities';

const contexts = (require as any).context('../../apps/api/src/migrations/', true, /\.ts$/);
const migrations = contexts.keys()
  .map(modulePath => contexts(modulePath))
  .reduce((result, migrationModule) => {
    result.concat(Object.keys(migrationModule).map(key => migrationModule[key]), [])
  });

export const dbConfig: TypeOrmModuleOptions = {
  type: 'postgres',
  host: process.env.POSTGRES_HOST,
  port: Number(process.env.POSTGRES_PORT),
  username: process.env.POSTGRES_USER,
  password: process.env.POSTGRES_PASSWORD,
  database: process.env.POSTGRES_DATABASE,
  entities: Object.values(entities),
  synchronize: false,
  migrationsRun: true,
  logging: true,
  migrations: Object.values(migrations)
};
```

**Comment by NestorPerez13** [comment]
I had the same issue, in my case was because i was exporting a function from the migration class:

`export class almacenes1604624108067 implements MigrationInterface {

  public async up(queryRunner: QueryRunner): Promise<void> {
    // as 2020-12-18 this warehouse's are not longer used.
    if (false) {
      const warehouses = ((await this.parseFile(await fs.readFile('./src/migrations/warehouse20201105.csv', 'utf-8'))).data) as Array<Warehouse>;
      var i, j, temparray, chunk = 500;
      for (i = 0, j = warehouses.length; i < j; i += chunk) {
        temparray = warehouses.slice(i, i + chunk);
        await getRepository(Warehouse).save(temparray).then();
      }
    }
  }

  public async down(queryRunner: QueryRunner): Promise<void> {
    // todo
  }

  async parseFile(file): Promise<ParseResult<any>> {
    return new Promise((resolve, reject) => {
      Papa.parse(file, {
        header: true,
        complete(results: ParseResult<any>, file?: File) {
          resolve(results);
        },
      });
    });
  }

}


export const parseFile = () => {
  
}`

I removed the exported function **parseFile**, and everything works ok now.

**Comment by bengotow** [comment]
Yep it looks like the ONLY export from a migration file must be the migration class. I had exported a helper function to use in a second migration and it caused this error!

**Comment by knnwulf** [comment]
> This is the correct way to initialize migrations and entities:-
> ([nrwl/nx#803](https://github.com/nrwl/nx/issues/803))
> 
> ```
> import * as entities from './src/entities';
> 
> const contexts = (require as any).context('../../apps/api/src/migrations/', true, /\.ts$/);
> const migrations = contexts.keys()
>   .map(modulePath => contexts(modulePath))
>   .reduce((result, migrationModule) => {
>     result.concat(Object.keys(migrationModule).map(key => migrationModule[key]), [])
>   });
> 
> export const dbConfig: TypeOrmModuleOptions = {
>   type: 'postgres',
>   host: process.env.POSTGRES_HOST,
>   port: Number(process.env.POSTGRES_PORT),
>   username: process.env.POSTGRES_USER,
>   password: process.env.POSTGRES_PASSWORD,
>   database: process.env.POSTGRES_DATABASE,
>   entities: Object.values(entities),
>   synchronize: false,
>   migrationsRun: true,
>   logging: true,
>   migrations: Object.values(migrations)
> };
> ```

I'm having more than 1 migration file in the directory and above code was not working in that case, because result.concat was failing (result is an object). So, I changed a little bit the code:

```js
const migrations = contexts.keys()
  .map(modulePath => {
    console.log('modulePath: ', modulePath);
    console.log('contexts: ', contexts(modulePath));
    return contexts(modulePath);
  }).reduce((result, migrationModule) => {
    return Object.assign(result, migrationModule);
  });
```

**Comment by ReallySmallSoftware** [comment]
I had this issue and I'm pulling my hair out! I ran the same migration locally, and also on my pre-production server and it worked fine. Fell over on production.

Turns out it didn't like the name of either my migration file, or the class within it (not sure which). I specify a list of migrations as an array in the config, not by having it find them in the directory automatically.

The migration was called UID1618341998674 when it failed, but it worked when I renamed it Uid1618341998674 - and changed all occurrences of UID to Uid.

I'm not too bothered it needed a rename. I am *very* bothered I ran it twice successfully and it failed at the critical time. I'm sure it will just be config or a version difference somewhere but right now I have no idea. :-(

**Comment by oliverlorenz** [comment]
Just as additional hint: Make sure in your migration array is not a `undefined`. I loaded my migrations that way:
```
export * from './1614354536191-migration1';
export * from './1617104377272-migration2';

export const migrations = Object.keys(module.exports).map(
	(key) => module.exports[key]
).filter((value) => !!value);
```
Solution was to filter this array afterwards.

**Comment by krekhovetskyi** [comment]
one more possible solution for initialization migrations and entities https://github.com/nrwl/nx/issues/803#issuecomment-864008312

**Comment by imnotjames** [comment]
I think we can catch this and make the error message more descriptive.

**Comment by vanduc1102** [comment]
I am having this issue with `require.context` 
```
TypeORMError:  migration name is wrong. Migration class name should have a JavaScript timestamp appended.
```
Does anyone know a way to fix ?

**Comment by dyamborko** [comment]
Try to delete all files from your migration folder or search for incorrect one. Because it can includes some broken migration or some other incorrect file.

**Comment by oli-clive-griffin** [comment]
I've just had this error and realised it was caused by exporting a function from a migration file (don't ask 😅), I imagine this interrupted the expected order of instantiating migration classes somehow.
